servicemanager/servicemanager.go

```diff
@@ -264,34 +368,142 @@ func (sm *ServiceManager) doApplyService(service ServiceInfo) (err error) {
 		}
+	}
+
+	service.Timestamp = time.Now().UTC()
+
+	if err = sm.serviceInfoProvider.ActivateService(service); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
+
+func (sm *ServiceManager) doRevertService(service ServiceInfo) error {
+	if _, err := sm.removeService(service); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	return nil
+}
 
-		if err := sm.serviceInfoProvider.RemoveService(oldService); err != nil {
-			log.Errorf("Can't remove old service from storage: %s", err)
+func (sm *ServiceManager) removeOutdateService(
+	availableSize, requiredSize int64,
+) (deallocatedLayersSize int64, err error) {
+	servicesInfo, err := sm.serviceInfoProvider.GetAllServices()
+	if err != nil {
+		return 0, aoserrors.Wrap(err)
+	}
+
+	ttl := sm.serviceTTLDays
+
+	for _, service := range servicesInfo {
+		if service.Timestamp.Add(time.Hour * 24 * time.Duration(ttl)).Before(time.Now()) {
+			deallocatedSize, err := sm.removeService(service)
+			if err != nil {
+				return 0, aoserrors.Wrap(err)
+			}
+
+			deallocatedLayersSize += deallocatedSize
+
+			if availableSize+deallocatedLayersSize > requiredSize {
+				return deallocatedLayersSize, nil
+			}
 		}
 	}
 
-	if err = sm.serviceInfoProvider.ActivateService(service); err != nil {
+	return 0, aoserrors.New("not enough memory")
+}
+
+func (sm *ServiceManager) doRemoveService(serviceID string) error {
+	service, err := sm.serviceInfoProvider.GetService(serviceID)
+	if err != nil {
 		return aoserrors.Wrap(err)
 	}
 
+	ttl := sm.serviceTTLDays
+
+	if service.Timestamp.Add(time.Hour * 24 * time.Duration(ttl)).Before(time.Now()) {
+		if _, err := sm.removeService(service); err != nil {
+			return aoserrors.Wrap(err)
+		}
+	} else {
+		sm.Lock()
+
+		sm.pendingRemove = append(sm.pendingRemove, service)
+
+		sm.Unlock()
+	}
+
 	return nil
 }
 
-func (sm *ServiceManager) doRevertService(service ServiceInfo) (retErr error) {
+func (sm *ServiceManager) removeService(service ServiceInfo) (layersSize int64, err error) {
+	imageParts, err := getImageParts(service.ImagePath)
+	if err != nil {
+		return 0, aoserrors.Wrap(err)
+	}
+
+	servicesInfo, err := sm.serviceInfoProvider.GetAllServices()
+	if err != nil {
+		return 0, aoserrors.Wrap(err)
+	}
+
+layersLoop:
+	for _, digest := range imageParts.LayersDigest {
+		for _, serviceInfo := range servicesInfo {
+			if serviceInfo.ServiceID == service.ServiceID && serviceInfo.AosVersion == service.AosVersion {
+				continue
+			}
+
+			serviceImageParts, err := getImageParts(serviceInfo.ImagePath)
+			if err != nil {
+				return 0, aoserrors.Wrap(err)
+			}
+
+			for _, compareDigest := range serviceImageParts.LayersDigest {
+				if digest == compareDigest {
+					continue layersLoop
+				}
+			}
+		}
+
+		size, err := sm.layerProvider.UninstallLayer(digest)
+		if err != nil {
+			return 0, aoserrors.Wrap(err)
+		}
+
+		log.Printf("deallocate service size: %v", size)
+		layersSize += size
+	}
+
+	sm.Lock()
+
+	if sm.isLayerSamePartition {
```

Comment:
`removeService` should not handle allocated size, it should return freed service size and freed layers size as separate params. sm.availableSize should be handled by caller.

Reply:
Done